translator/trace/protospan_to_jaegerthrift.go

```
@@ -0,0 +1,321 @@
+	}
+
+	if tagValue != "" {
+		jTag := &jaeger.Tag{
+			Key:  "span.kind",
+			VStr: &tagValue,
+		}
+		jTags = append(jTags, jTag)
+	}
+
+	return jTags
+}
+
+func ocTimeEventsToJaegerLogs(ocSpanTimeEvents *tracepb.Span_TimeEvents) []*jaeger.Log {
+	if ocSpanTimeEvents == nil || ocSpanTimeEvents.TimeEvent == nil {
+		return nil
+	}
+
+	ocTimeEvents := ocSpanTimeEvents.TimeEvent
+
+	// Assume that in general no time events are going to produce nil Jaeger logs.
+	jLogs := make([]*jaeger.Log, 0, len(ocTimeEvents))
+	for _, ocTimeEvent := range ocTimeEvents {
+		jLog := &jaeger.Log{
+			Timestamp: timestampToEpochMicroseconds(ocTimeEvent.Time),
+		}
+		switch teValue := ocTimeEvent.Value.(type) {
+		case *tracepb.Span_TimeEvent_Annotation_:
+			jLog.Fields = ocAnnotationToJagerTags(teValue.Annotation)
+		case *tracepb.Span_TimeEvent_MessageEvent_:
+			jLog.Fields = ocMessageEventToJaegerTags(teValue.MessageEvent)
+		default:
+			msg := "An unknown OpenCensus TimeEvent type was detected when translating to Jaeger"
+			jTag := &jaeger.Tag{
+				Key:  "unknown.oc.timeevent.type",
+				VStr: &msg,
+			}
+			jLog.Fields = append(jLog.Fields, jTag)
+		}
+
+		jLogs = append(jLogs, jLog)
+	}
+
+	return jLogs
+}
+
+func ocAnnotationToJagerTags(annotation *tracepb.Span_TimeEvent_Annotation) []*jaeger.Tag {
+	if annotation == nil {
+		return nil
+	}
+
+	// TODO: (@pjanotti) what about Description? Does it fit as another tag?
+
+	return ocSpanAttributesToJaegerTags(annotation.Attributes)
+}
+
+func ocMessageEventToJaegerTags(msgEvent *tracepb.Span_TimeEvent_MessageEvent) []*jaeger.Tag {
+	if msgEvent == nil {
+		return nil
+	}
+
+	// TODO: (@pjanotti) Not clear how to map those to Jaeger, perhaps some OpenTracing tags...
+
+	return nil
+}
+
+func truncableStringToStr(ts *tracepb.TruncatableString) string {
+	if ts == nil {
+		return ""
+	}
+	return ts.Value
+}
+
+func traceIDBytesToLowAndHigh(traceID []byte) (traceIDLow, traceIDHigh int64) {
+	if len(traceID) != 16 {
+		return
```

It would be useful to return an error instead of silently passing as TraceID is a very important field.